### PR TITLE
NL: Merge Custom + Base indexes on the fly

### PR DESCRIPTION
**Pull request by pradh**
Builds on top of https://github.com/datacommonsorg/website/pull/4130

Currently, the custom_ft + medium_ft embeddings are merged on server load.

Instead, we merge in custom_ft index in the query path, when the env is custom DC (and the requested index is not custom_ft already).

Why?

* custom_ft can merge with any index besides medium_ft, based on the `idx=` setting
* Allows interoperability between different custom DC vs. base DC embedding stores.  For example, custom DC might be in-memory while base DC can be LanceDB or Vertex AI based.
* This will also allow unblocking UN work.  So when we use `dc=undata`, which implies `idx=undata_ft`, the stuff in `custom_ft` will get merged in

### CODE REVIEW

File: nl_server/tests/merge_test.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+              EmbeddingsMatch(sentence='poverty',
+                              score=0.85,
+                              vars=['WHO_Poverty']),
+              EmbeddingsMatch(sentence='above poverty',
+                              score=0.7,
+                              vars=['WHO_AbovePoverty']),
+          ]
+      }, {
+          'not poor': [
+              EmbeddingsMatch(sentence='below poverty',
```
Comment (keyurva) [inline nl_server/tests/merge_test.py:38]: This is an edge case but do we want to address a scenario where the same SV is in both inputs?

i.e. the same embeddings match (say `EmbeddingsMatch('below poverty', 0.9, ['Count_BelowPovertyLevel'])` exists in both? Currently the output list will return this match twice. Do we want the merge function to account for duplicates?

File: nl_server/util.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+# unless required by applicable law or agreed to in writing, software
+# distributed under the license is distributed on an "as is" basis,
+# without warranties or conditions of any kind, either express or implied.
+# see the license for the specific language governing permissions and
+# limitations under the license.
+
+import os
+
+
+def is_custom_dc() -> bool:
```
Comment (keyurva) [inline nl_server/util.py:18]: Can we move this under `shared.lib`?

**Review by keyurva — COMMENTED** [review]
Thanks - this is a great simplification and will help immensely moving forward!

File: nl_server/merge.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+  # Group by query first.
+  query_grouping: Dict[str, List[EmbeddingsResult]] = {}
+  for input in inputs:
+    for query, emb_result in input.items():
+      if query not in query_grouping:
+        query_grouping[query] = []
+      query_grouping[query].append(emb_result)
+
+  # Merge each query group.
+  result: List[SearchVarsResult] = {}
```
[the remark this answers is not in the document]
Comment (pradh) [reply nl_server/merge.py:33]: Good catch

**Review by chejennifer — APPROVED** [review]
nice!

File: nl_server/merge.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+  for query, emb_result in query_grouping.items():
+    result[query] = _merge_search_results_for_one_query(emb_result)
+
+  return result
+
+
+def _merge_search_results_for_one_query(
```
[the remark this answers is not in the document]
Comment (pradh) [reply nl_server/merge.py:40]: Great suggestion, agree on the perf vs. simplicity tradeoff.

File: nl_server/search.py [hunk 5]
```diff
@@ -31,23 +32,38 @@
...
   topk = _get_topk(skip_topics)
 
-  query2candidates = embeddings.vector_search(queries, topk)
+  # Call vector search for each index.
+  query2candidates_list: List[EmbeddingsResult] = []
+  for embeddings in embeddings_list:
+    query2candidates_list.append(embeddings.vector_search(queries, topk))
+
+  # Merge the results.
+  if len(query2candidates_list) == 1:
```
[the remark this answers is not in the document]
Comment (pradh) [reply nl_server/search.py:51]: Done.

File: nl_server/tests/merge_test.py (hunk 1 above)
Comment (pradh) [reply nl_server/tests/merge_test.py:38]: There are a few edge cases:
1.  Same sentence appears in custom DC and base DC with different SVs
2. Same variable appears in custom DC and base DC with different sentences
3. Same sentence + variable appear

`_rank_vars` will group by SVs and rerank them, and should handle (1) + (2).  (3) is a good point, I've handled that.

File: nl_server/util.py (hunk 2 above)
Comment (pradh) [reply nl_server/util.py:18]: `shared` gets used in many contexts (not just NL and Web server), so I think its better to keep this here?

File: nl_server/loader.py [hunk 6]
```diff
@@ -92,7 +92,7 @@ def load_custom_embeddings(app: Flask):
   # This is intentional.
   nl_embeddings: emb_map.EmbeddingsMap = app.config[config.NL_EMBEDDINGS_KEY]
   # Merge custom index with default embeddings.
```
[the remark this answers is not in the document]
Comment (pradh) [reply nl_server/loader.py:94]: Good point, I needed to update the comments.

**Review by pradh — COMMENTED** [review]
Thanks for the review!

**Review by keyurva — APPROVED** [review]
Thanks for the changes!
